sdlib/d/gc/slab.d

```diff
@@ -23,6 +25,228 @@ struct SlabAllocGeometry {
 	}
 }
 
+struct SlabAllocInfo {
+private:
+	SlabAllocGeometry sg;
+	Extent* e;
+	bool scSupportsMetaData = false;
+	bool metaDataEnabled = false;
+
+public:
+	@property
+	auto allowsMetaData() {
+		return scSupportsMetaData;
+	}
+
+	@property
+	auto hasMetaData() {
+		return metaDataEnabled;
+	}
+
+	this(PageDescriptor pd, const void* ptr) {
+		assert(pd.isSlab(), "Expected a slab!");
+
+		e = pd.extent;
+		sg = SlabAllocGeometry(pd, ptr);
+		scSupportsMetaData = sizeClassSupportsMetadata(pd.sizeClass);
+		metaDataEnabled = scSupportsMetaData && e.hasMetaData(sg.index);
+	}
+
+	@property
+	auto address() {
+		return sg.address;
+	}
+
+	@property
+	size_t slotCapacity() {
+		return sg.size;
+	}
+
+	@property
+	size_t usedCapacity() {
+		return sg.size - freeSpace;
+	}
+
+	bool setUsedCapacity(size_t size) {
+		if (!scSupportsMetaData || (size > slotCapacity)) {
+			return false;
+		}
+
+		setFreeSpace(sg.size - size);
+		return true;
+	}
+
+	@property
+	size_t freeSpace() {
+		return metaDataEnabled ? readPackedFreeSpace(freeSpacePtr) : 0;
+	}
+
+	void setFreeSpace(size_t size) {
+		assert(scSupportsMetaData, "size class not supports slab metadata!");
+		assert(size <= sg.size, "size exceeds alloc size!");
+
+		if (size == 0) {
+			disableMetaData();
+			return;
+		}
+
+		writePackedFreeSpace(freeSpacePtr, size & ushort.max);
+		enableMetaData();
+	}
+
+private:
+	void enableMetaData() {
+		assert(scSupportsMetaData, "size class not supports slab metadata!");
+
+		if (!metaDataEnabled) {
+			e.enableMetaData(sg.index);
+			metaDataEnabled = true;
+		}
+	}
+
+	void disableMetaData() {
+		assert(scSupportsMetaData, "size class not supports slab metadata!");
+
+		if (metaDataEnabled) {
+			e.disableMetaData(sg.index);
+			metaDataEnabled = false;
+		}
+	}
+
+	@property
+	T* ptrToAllocEnd(T)() {
+		return cast(T*) (sg.address + sg.size) - T.sizeof;
+	}
+
+	alias freeSpacePtr = ptrToAllocEnd!ushort;
+}
+
+unittest SlabAllocInfo {
+	import d.gc.base;
+	static Base base;
+	scope(exit) base.clear();
+
+	auto slot = base.allocSlot();
+	auto e = Extent.fromSlot(0, slot);
+	auto block = base.reserveAddressSpace(HugePageSize);
+	assert(block !is null);
+
+	static SlabAllocInfo simulateSmallAlloc(size_t size) {
+		auto ec = ExtentClass.slab(getSizeClass(size));
+		e.at(block, PageSize, null, ec);
+		assert(e.allocate() == 0);
```

Comment:
We are only testing slot 0. We had a bug with that in the past, a strong tell that we probably should test different slots.